### ISSUE TITLE
Cookie banner not saving

### ISSUE THREAD

**Issue by gokkstudio** [fabiodalez-dev/FAZ-Cookie-Manager#125]
Am I doing something wrong or cookie banner is not saving changes?
Added new banner, still not saving. Clues?

**Comment by fabiodalez-dev** [comment]
Hi @gokkstudio — thanks for the report, sorry you're hitting this. To help diagnose I need a bit more detail since "not saving" can have several different shapes:

1. **What exactly happens when you click Save?**
   - Does a success message ("Saved" / green toast) appear, but the settings revert when you reload the page?
   - Does an error message appear?
   - Does the Save button do nothing at all (no spinner, no toast, no error)?
   - Does the page reload but the changes are gone?

2. **Plugin and WordPress versions:**
   - Plugin version: `wp-admin → Plugins` page (or `Tools → Site Health → Info → Plugins`)
   - WordPress version: bottom-right corner of any wp-admin page
   - PHP version: `Tools → Site Health → Info → Server`

3. **Browser DevTools — open the console (F12) BEFORE clicking Save:**
   - Any red errors in the **Console** tab when you load the banner edit page?
   - Any red errors when you click Save?
   - In the **Network** tab, does the Save click trigger a request to `/wp-json/faz/v1/banners/...`? What HTTP status does it return (200? 403? 500?)?
   - A screenshot of the Console + Network tab around the Save click would be very helpful.

4. **Other active plugins:** can you list them? Caching plugins (WP Rocket, W3 Total Cache, LiteSpeed, etc.), security plugins (Wordfence, Sucuri), or "REST API hardening" plugins are the most common interferers.

5. **Active theme.**

6. **Did this work before, or has it always failed?** If it worked before — what changed (plugin update, theme update, WP update, new plugin installed)?

Once I have at least items 1, 3 (the Network status code), and 4, I can start narrowing down. If it's reproducible on a fresh WordPress install with only FAZ Cookie Manager active, that points to a plugin bug; if it only fails on your live site with other plugins active, it's almost certainly a conflict and we can identify which one.

— Fabio

**Comment by nathanchenly** [comment]
I'm getting the same issue of not able to create or save the banner.
Console shows this error: _Blocked script execution in 'https://example.com/?faz_banner_preview=1' because the document's frame is sandboxed and the 'allow-scripts' permission is not set._

**Comment by fabiodalez-dev** [comment]
Hi @nathanchenly, @gokkstudio,

I want to be completely honest with you: I have not been able to reproduce this, and I've genuinely tried.

On a clean WordPress install with the latest FAZ Cookie Manager as the only active plugin, I created and edited a banner, changed settings, clicked **Save**, and the changes persisted correctly after a reload. The save request (`PUT /wp-json/faz/v1/banners/{id}`) returned **200** and the data was written. So saving is not broken in the plugin on its own — which tells me something specific to *your* environment is getting in the way, and I simply can't see it from here.

About the console error you pasted — *"Blocked script execution in '…/?faz_banner_preview=1' because the document's frame is sandboxed"* — I dug into it carefully, and I'm fairly confident it's a **red herring**, not the cause of the save failing. That URL is the live banner preview, which FAZ loads inside a sandboxed iframe. The sandbox deliberately blocks the *previewed page's own scripts* (your theme, other plugins, analytics) so they don't run inside the admin. The banner is drawn into that iframe by the admin page directly, not by those blocked scripts, and the **Save** button is a separate REST request that never touches the iframe. So that message is just noise from your site's scripts being blocked in the preview — it does not stop the save. (The `example.com` in your report is only the placeholder you put in place of your real domain; FAZ always uses your actual site URL there.)

Everything points to an **environment-specific issue** — most likely a conflict with another plugin or a server-side rule, not a FAZ bug in isolation. The usual suspects:

- a security / "REST API hardening" plugin (Wordfence, Solid Security, "Disable REST API", etc.) blocking authenticated REST writes;
- a WAF / ModSecurity rule or reverse proxy blocking `PUT` requests or JSON bodies;
- a caching/optimization layer serving a stale admin nonce.

To actually pin this down I need to see what your setup does that mine doesn't. Could you help me with the part that matters most:

1. Open the banner edit page → press **F12 → Network** → click **Save**, and tell me the status code of the request to `/wp-json/faz/v1/banners/…` — is it **200, 403, 401, 500**, or does it not fire at all? A screenshot of that request (including the **Response** tab) is ideal.
2. Your full list of **active plugins** and your **active theme**.
3. Your server stack if you know it (host, PHP version, whether Cloudflare or a WAF sits in front).

If it's easier, the fastest path for me is temporary access to a **staging copy** of the site — or even a **System Status export** (FAZ → System Status) — so I can drive the exact failing save myself in your environment. I'm glad to debug it directly; I just can't fix what I can't reproduce, and right now everything I'm able to test passes.

Thanks for your patience, and sorry for the trouble. Even just the Network status code will already tell me a great deal about where it's breaking.

— Fabio

**Comment by gokkstudio** [comment]
Hey Guys, sorry for being away and not providing the necessary data to properly do the debugging.
As far as I can tell, the original post was related to a certain environment like Fabio mentioned - in my case: CloudFlare proxied domain with bunch of WAF rules, dedicated Nginx server tailored for single client only etc... I really had no time to dig deeper and used another solution.

But, I used FAZ again, on another domain, my standard Apache config on cPanel/WHM server with no issues at all.
So the problem must be tied to environment.
@nathanchenly can you give us your stack so I can compare it to mine please?

I think we can find out what is messing with it pretty quickly.

**Comment by setsailfor** [comment]
I'm unfortunately also having this issue. Thought it might be Safari or AdGuard, but that appears to not be the case. I think I got the request info you were asking the other person for, let me know if I misunderstood. Looks like the code is 200. Also attaching a system status export as a text file.

[fazcookie-systemexport-ybw.rtf](https://github.com/user-attachments/files/29873724/fazcookie-systemexport-ybw.rtf)

<img width="1430" height="1019" alt="Image" src="https://github.com/user-attachments/assets/6e79d0ff-fa4c-4719-9e6c-5391abe6f9dc" />
<img width="1430" height="1019" alt="Image" src="https://github.com/user-attachments/assets/10eb8b81-eb29-40d6-8690-7d248dc8ea8f" />

I'm running Woocommerce on a WordPress.com site. I don't know what my stack is otherwise, I'm sorry. If you let me know where to look, I can poke around and maybe find out. Installed plugins: Akismet Anti-spam, Classic Editor (not enabled), FAZ Cookie Manager, Gutenberg, Jetpack, Layout Grid (not enabled), Page Optimize, Stripe Tax, WooCommerce, WooCommerce Stripe Gateway.

Let me know if anything else would be helpful!

**Comment by setsailfor** [comment]
After I commented, it appears one of the changes I was trying to make (choosing a different design preset) stuck. Trying to change the wording and clicking Save still doesn't apply for me.

**Comment by fabiodalez-dev** [comment]
Hi @setsailfor, thanks for the information — the system status export and the two Network screenshots already tell me something important.

Looking at your screenshots, when you change the wording and click Save the only requests that fire are a GET for cookie.png and one POST to admin-ajax.php (initiated by jQuery, almost certainly Jetpack or WordPress.com heartbeat, not my plugin). What is NOT there is the request that actually saves the banner: a POST to /wp-json/faz/v1/banners/…. So the 200 you saw is the cookie.png image, not the save. This isn't a case of "it saves and then reverts" — the save request is never being sent at all. That's why your text change doesn't stick, while a design preset happened to save at some earlier point (there are 3 banner rows in your database, so saving does work on your site sometimes).

The reason the save request never fires is almost certainly the 3 errors I can see in your DevTools (the red "3" in the toolbar). My admin script attaches the Save button's behaviour when the banner editor loads; if that script throws an error during load, the button becomes a dead button — clicking it does nothing, no request goes out, no error toast appears. That matches your symptom exactly.

So the one thing that would crack this open is the Console tab, not the Network tab. Could you open the banner page, press F12, click the Console tab (not Network), reload the page, and send me a screenshot of those 3 red errors? That will most likely name the exact file and line that's breaking.

To be completely honest: I tried to reproduce this by installing Page Optimize on a clean WordPress 7 / PHP 8.4 site and saving a banner, and there it kept working — the save fired and the text persisted. So I can't yet point the finger at one specific plugin from my side; your WordPress.com environment (Jetpack, Page Optimize, the standalone Gutenberg plugin, plus WordPress.com's own admin scripts) is doing something a normal install doesn't, and the console errors are what will show me what.

While you grab that, the fastest test on your end is to rule the interference out directly: temporarily deactivate Page Optimize, then retry the text save; if it still fails, reactivate it and try the same with Jetpack, then with the Gutenberg plugin (you're on WordPress 7, which already has the block editor built in, so the separate Gutenberg plugin is a common source of admin-script conflicts). Whichever one makes Save start working is the culprit, and I can then add a proper compatibility fix on my side.

One note for context: this is a different problem from the one @gokkstudio originally hit — his was a server/WAF layer (Cloudflare) blocking the request before it reached WordPress. Yours is the opposite end: the request isn't even leaving the browser. Same visible symptom, different cause.

Thanks for the detailed report and the patience — the console screenshot should get us to a fix quickly.

— Fabio

**Comment by joxerg** [comment]
Same problem here. Using Wordpress in a VPS with Docker, maybe some cache issue? using Flyingpress...

Keep the good job!!

**Comment by fabiodalez-dev** [comment]
Hi @joxerg, thanks for the report. FlyingPress is the key detail — like the earlier case in this thread, it's an asset optimizer, and the most likely cause is that it's minifying / combining / deferring my admin JavaScript inside wp-admin. When that happens, the script that wires up the Save button never runs, so clicking Save sends no request at all — the banner is never saved and no error appears.

To pin it down I need one of these:

1. **The Console errors.** Open the Cookie Banner page, press F12, click the **Console** tab (not Network), reload, click Save, and send me a screenshot of any red errors — that usually names the exact file that's breaking.
2. **A quick self-test.** In FlyingPress, disable JS optimization / minify / defer for the admin area (or exclude my plugin's scripts), clear its cache, and retry Save. If it starts working, FlyingPress is the culprit and I'll ship a compatibility fix.
3. **Direct access.** If it's easier, send me a link to the site (or temporary admin access) and I'll look myself.

One note: since the save request never leaves the browser, this is a JS-optimization issue, not a page-cache one — so FlyingPress's caching isn't the suspect, its JS handling is.

— Fabio

**Comment by joxerg** [comment]
Hi, i managed to save after:
_ deactivating Redis cache plugin
_ deactivating (Flyingpress and purging cache) plugin
_ activating Cache compatibility mode

I think main issue is Flyingpress here. Thank you!!

**Comment by fabiodalez-dev** [comment]
Thanks @joxerg — your steps actually pinpointed **two** separate real issues on my side, and I'm working on a fix for both.

1. **Redis.** With a persistent object cache (Redis Object Cache, Memcached), my banner-config cache wasn't fully invalidated on save. My cache-clear scanned `wp_options` for transients, but under a persistent object cache transients don't live in `wp_options` — so the stale banner payload survived in Redis and kept being served back to the editor and the frontend. That's exactly the "it saves in the DB but the old version keeps showing" symptom, and why you had to deactivate the Redis plugin, not just FlyingPress. Fixed by rotating the cache-key seed so stale entries become unreachable on both backends.

2. **FlyingPress.** FlyingPress is stricter than the other cache plugins I already support: it does **not** honor `DONOTCACHEPAGE` or `Cache-Control: no-store`, which is how I tell every other cache "don't cache this page" — so my existing page-cache bypass was silently inert under FlyingPress. I've added FlyingPress-specific handling: an automatic purge (+ re-warm) on every banner/cookie/settings save, plus its documented `flying_press_is_cacheable` filter to skip caching on pages whose output varies by country.

Cache Compatibility Mode is the right stopgap for now; these fixes will remove the need for it. I'm still going through FlyingPress's docs for a cleaner hook on the JS-delay side (excluding my admin scripts programmatically) and will follow up here. Thanks for the detailed report.

— Fabio

**Comment by setsailfor** [comment]
Hi! I'm sorry for the late reply. Attaching screenshots of the Console view! Let me know if there's something I didn't catch that would be relevant to you. However, good news, temporarily disabling Gutenberg did the trick!

<img width="1317" height="494" alt="Image" src="https://github.com/user-attachments/assets/e889d87f-2cae-4ee6-9d2f-a3b19ec01d99" />
<img width="1318" height="796" alt="Image" src="https://github.com/user-attachments/assets/8dcfe514-3a7d-4952-b4cc-f79d100f6bad" />

**Comment by fabiodalez-dev** [comment]
Following up @joxerg — both fixes from your report are done and ship in the next release.

Redis / persistent object cache: the banner-config cache is now invalidated by rotating its cache-key seed, so a save makes the old entry unreachable on both backends — the DB transient AND Redis/Memcached. The stale banner no longer survives in Redis after a save, so you won't need to deactivate the Redis plugin.

FlyingPress: three parts, since FlyingPress doesn't honor DONOTCACHEPAGE or Cache-Control: no-store the way the other caches I support do:

1. It now purges FlyingPress automatically on every banner / cookie / settings save. The purge is HTML-only, so a normal save doesn't kick off a full-site preload crawl.
2. It hooks FlyingPress's documented flying_press_is_cacheable filter, so pages whose banner varies by country are skipped instead of frozen for everyone.
3. The consent scripts are now auto-excluded from FlyingPress's "Delay JavaScript" and minify (both FlyingPress 4 and 5), so the banner is never held back until the first interaction — this is the JS-delay side I said I was still working on.

Once this release is out you won't need Cache Compatibility Mode for FlyingPress or Redis; it goes back to being optional. I'll comment here when it's published. Thanks again — your steps are what pinpointed both issues.

**Comment by fabiodalez-dev** [comment]
Thanks for the console screenshots — they make it clear, and the good news is that none of those errors come from my plugin.

Reading them: every red error traces back to the standalone Gutenberg plugin (its bundled react-dom 18.3.1 under /wp-content/plugins/gutenberg/) and to WordPress.com's own admin scripts (help-center-wp-admin.min.js, wc-calypso-bridge), plus two "Failed to load script" errors for stats.wp.com/w.js and nostats.js that your ad blocker (AdGuard) is blocking. Those uncaught errors fire while the admin page is still loading, and that is what leaves my Save button unwired — exactly the dead-button symptom from before, where the click sends no request at all.

Disabling Gutenberg was the right call. On WordPress 7 the block editor is already built into core, so the separate Gutenberg plugin is redundant here, and because it ships its own copy of React it clashes with the one WordPress already loads — a well-known source of admin-script breakage. Keeping it deactivated is safe; you lose nothing on WP 7.

So there is nothing to fix on your data side — your saves were fine, the button just could not send them. Thanks for taking the time to capture the console; that is what confirmed it.
